src/lease-manager.ts

```diff
@@ -0,0 +1,209 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {freemem} from 'os';
+import * as defer from 'p-defer';
+
+import {Message} from './message-stream';
+import {Subscriber} from './subscriber';
+
+/**
+ * @typedef {object} FlowControlOptions
+ * @property {number} [maxBytes] The maximum amount of memory to allow message
+ *     data to consume. Defaults to 20% of available memory.
+ * @property {number} [maxExtension=Infinity] The maximum duration to extend
+ *     the message deadline before re-delivering.
+ * @property {number} [maxMessages=100] The maximum number of messages to allow
+ *     in memory before pausing the message stream.
+ */
+export interface FlowControlOptions {
+  maxBytes?: number;
+  maxExtension?: number;
+  maxMessages?: number;
+}
+
+/**
+ * Manages a Subscribers inventory while auto-magically extending the message
+ * deadlines.
+ *
+ * @private
+ * @class
+ *
+ * @param {Subscriber} sub The subscriber to manage leases for.
+ * @param {FlowControlOptions} options Flow control options.
+ */
+export class LeaseManager {
+  bytes: number;
+  _isLeasing: boolean;
+  _messages: Set<Message>;
+  _onfree?: defer.DeferredPromise<void>;
+  _options!: FlowControlOptions;
+  _subscriber: Subscriber;
+  _timer?: NodeJS.Timer;
+  constructor(sub: Subscriber, options = {}) {
+    this.bytes = 0;
+    this._isLeasing = false;
+    this._messages = new Set();
+    this._subscriber = sub;
+
+    this.setOptions(options);
+  }
+  /**
+   * @type {number}
+   */
+  get size(): number {
+    return this._messages.size;
+  }
+  /**
+   * Adds a message to the inventory, kicking off the auto-extender if need be.
+   *
+   * @param {Message} message The message.
+   */
+  add(message: Message): void {
+    this._messages.add(message);
+    this.bytes += message.length;
+
+    if (!this._isLeasing) {
+      this._isLeasing = true;
+      this._scheduleExtension();
+    }
+  }
+  /**
+   * Removes ALL messages from inventory.
+   */
+  clear(): void {
+    this._cancelExtension();
+    this._messages.clear();
+    this.bytes = 0;
+  }
+  /**
+   * Indicates if we're at/over capacity or not.
+   *
+   * @returns {boolean}
+   */
+  isFull(): boolean {
+    const {maxBytes, maxMessages} = this._options;
+    return this.size >= maxMessages! || this.bytes >= maxBytes!;
+  }
+  /**
+   *
+   */
+  onFree(): Promise<void> {
+    if (!this._onfree) {
+      this._onfree = defer();
```

Comment:
We definitely can do that if you feel strongly about it, I went with the [p-defer](https://github.com/sindresorhus/p-defer) module because its tiny and tackles the problem in the same way I would have.

Reply:
I guess I'm saying I don't understand the problem :) Does this just do a delay before executing a promise?

Reply:
Whoops, my bad! So there's a limit on how many messages we want in memory at a given time and I was using this promise as a way to determine when it was safe to introduce more. Initially I was going to go the EventEmitter route but felt like this was simpler solution.